### PR TITLE
Parameterize DeriveSha with block number

### PR DESCRIPTION
**Pull request by blukat29**
## Proposed changes

- Add `blockNum` argument to `types.DeriveSha()` and `types.EmptyRootHash()`.
  ```
  As-is: InitDeriveSha(type) + DeriveSha(list) + EmptyRootHash
  To-be: InitDeriveSha(chainConfig) + DeriveSha(list, num) + EmptyRootHash(num)
  ```
- The `num` argument is ignored in this PR.
- Hardfork and/or Governance logic will be added in a later PR
- `blockchain.InitDeriveSha` is relocated from `inside NewBlockChain()` to `cn.New()`
  - This is to feed the up-to-date `chainConfig` which is modified by `governance.UpdateParams()`.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

#1687

## Further comments

### CODE REVIEW

**Comment by blukat29** [comment]
Will be rebased after https://github.com/klaytn/klaytn/pull/1689 is merged.

File: node/cn/handler.go [hunk 1]
```diff
@@ -998,7 +998,7 @@ func handleReceiptsRequestMsg(pm *ProtocolManager, p Peer, msg p2p.Msg) error {
 		// Retrieve the requested block's receipts, skipping if unknown to us
 		results := pm.blockchain.GetReceiptsByBlockHash(hash)
 		if results == nil {
-			if header := pm.blockchain.GetHeaderByHash(hash); header == nil || header.ReceiptHash != types.EmptyRootHash {
+			if header := pm.blockchain.GetHeaderByHash(hash); header == nil || !header.EmptyReceipts() {
```
Comment (kjhman21) [inline node/cn/handler.go:1001]: it's little bit weird that EmptyReceipts() returns true if the receipthash == EmptyHash.
What about 'IsEmptyReceipts' or 'IsReceiptEmpty'?

Comment (blukat29) [reply node/cn/handler.go:1001]: This `func EmptyXX() bool` functions are used in several data structures in Klaytn codebase (e.g. `StateDB.Empty()`, `txList.Empty()`). If it's not too confusing, I would leave it as is.

Comment (kjhman21) [reply node/cn/handler.go:1001]: @blukat29 Okay. got it.

File: blockchain/types/derivesha/mux.go [hunk 2]
```diff
@@ -21,41 +21,59 @@
...
-	return deriveShaObj.DeriveSha(list)
+func DeriveShaMux(list types.DerivableList, num *big.Int) common.Hash {
+	return instances[getType(num)].DeriveSha(list)
+}
+
+func EmptyRootHashMux(num *big.Int) common.Hash {
+	return emptyRoots[getType(num)]
+}
+
+func getType(num *big.Int) int {
```
Comment (kjhman21) [inline blockchain/types/derivesha/mux.go:71]: Why do we receive `num` here?

Comment (blukat29) [reply blockchain/types/derivesha/mux.go:71]: It's unused right now but will be used in #1708.
I'll make it `getType()` for clarity, and then change it in #1708.

Comment (blukat29) [reply blockchain/types/derivesha/mux.go]: Fixed in cb91cae.

**Review by mckim19 — APPROVED** [review]

**Review by ian0371 — APPROVED** [review]

**Review by kjhman21 — APPROVED** [review]

**Review by hyunsooda — APPROVED** [review]

**Review by JayChoi1736 — APPROVED** [review]

**Review by henry-will — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]

**Pull request merged**